loopy/target/pyopencl.py

```diff
@@ -270,6 +272,197 @@ def dtype_to_ctype(self, dtype):
+
+        if n_complex and not d_complex:
+            return p.Variable("%s_divider" % self.complex_type_name(tgt_dtype))(
+                    self.rec(expr.numerator, type_context, tgt_dtype),
+                    self.rec(expr.denominator, type_context))
+        elif not n_complex and d_complex:
+            return p.Variable("%s_rdivide" % self.complex_type_name(tgt_dtype))(
+                    self.rec(expr.numerator, type_context),
+                    self.rec(expr.denominator, type_context, tgt_dtype))
+        else:
+            return p.Variable("%s_divide" % self.complex_type_name(tgt_dtype))(
+                    self.rec(expr.numerator, type_context, tgt_dtype),
+                    self.rec(expr.denominator, type_context, tgt_dtype))
+
+    def map_constant(self, expr, type_context):
+        if isinstance(expr, (complex, np.complexfloating)):
+            try:
+                dtype = expr.dtype
+            except AttributeError:
+                # (COMPLEX_GUESS_LOGIC) This made it through type 'guessing' in
+                # type inference, and it was concluded there (search for
+                # COMPLEX_GUESS_LOGIC in loopy.type_inference), that no
+                # accuracy was lost by using single precision.
+                cast_type = "cfloat"
+            else:
+                if dtype == np.complex128:
+                    cast_type = "cdouble"
+                elif dtype == np.complex64:
+                    cast_type = "cfloat"
+                else:
+                    raise RuntimeError("unsupported complex type in expression "
+                            "generation: %s" % type(expr))
+
+            return p.Variable("%s_new" % cast_type)(expr.real, expr.imag)
+
+        return super().map_constant(expr, type_context)
+
+    def map_power(self, expr, type_context):
+        tgt_dtype = self.infer_type(expr)
+        base_dtype = self.infer_type(expr.base)
+        exponent_dtype = self.infer_type(expr.exponent)
+
+        if not self.allow_complex or (not tgt_dtype.is_complex()):
+            return super().map_power(expr, type_context)
+
+        if expr.exponent in [2, 3, 4]:
+            value = expr.base
+            for i in range(expr.exponent-1):
+                value = value * expr.base
+            return self.rec(value, type_context)
+        else:
+            b_complex = base_dtype.is_complex()
+            e_complex = exponent_dtype.is_complex()
+
+            if b_complex and not e_complex:
+                return p.Variable("%s_powr" % self.complex_type_name(tgt_dtype))(
+                        self.rec(expr.base, type_context, tgt_dtype),
+                        self.rec(expr.exponent, type_context))
+            else:
+                return p.Variable("%s_pow" % self.complex_type_name(tgt_dtype))(
+                        self.rec(expr.base, type_context, tgt_dtype),
+                        self.rec(expr.exponent, type_context, tgt_dtype))
+
+# }}}
```

Comment:
I'm guessing this was mostly moved (which it should have been, long ago). Were there any functional changes?

Reply:
Nope, no functional changes here.